clippy_lints/src/casts/ptr_as_ptr.rs

```diff
@@ -3,12 +3,29 @@ use clippy_utils::diagnostics::span_lint_and_sugg;
 use clippy_utils::source::snippet_with_applicability;
 use clippy_utils::sugg::Sugg;
 use rustc_errors::Applicability;
-use rustc_hir::{Expr, ExprKind, Mutability, TyKind};
+use rustc_hir::{Expr, ExprKind, Mutability, QPath, TyKind};
+use rustc_hir_pretty::qpath_to_string;
 use rustc_lint::LateContext;
 use rustc_middle::ty::{self, TypeAndMut};
+use rustc_span::sym;
 
 use super::PTR_AS_PTR;
 
+enum OmitFollowedCastReason<'a> {
+    None,
+    Null(&'a QPath<'a>),
+    NullMut(&'a QPath<'a>),
```

Comment:
Note for future improvement:
This should be replaced with check of:
1. if calling function is `<T>(...args) -> *{const,mut} T`
2. one of:
	* `args` does not contain or refer `T` -> we can give arbitrary `T`
	* check if it still pass type-checking if we change `T=Src` to `T=Dst`